casper/src/main/scala/coop/rchain/casper/api/BlockApiImpl.scala

```diff
@@ -175,6 +178,55 @@ class BlockApiImpl[F[_]: Concurrent: RuntimeManager: BlockDagStorage: BlockStore
       )
   }
 
+  override def deployStatus(deployId: DeployId): F[ApiErr[DeployExecStatus]] = {
+
+    def getSeqPar(blockHash: String): F[Seq[Par]] =
+      for {
+        par       <- Sync[F].delay(GDeployId(deployId).asInstanceOf[Par])
+        dataAtPar <- getDataAtPar(par, blockHash, usePreStateHash = false)
+        seqPar    <- liftToBlockApiErr(dataAtPar).map { case (seqPar, _) => seqPar }
+      } yield seqPar
+
+    def withBlockInfoAndSeqPar[A](f: (BlockInfo, Seq[Par]) => A): F[A] =
+      for {
+        blockHash <- BlockDagStorage[F].lookupByDeployId(deployId).map(_.get.toHexString)
+        blockInfo <- getBlock(blockHash) >>= liftToBlockApiErr
+        seqPar    <- getSeqPar(blockHash)
+      } yield f(blockInfo, seqPar)
+
+    def notProcessed(message: String) = DeployExecStatus().withNotProcessed(NotProcessed(message))
+
+    for {
+      status <- BlockDagStorage[F].deployStatus(deployId)
```

Comment:
`deployStatus` that Anton created was just as an template how to get specific data and not to be used in its current form. Now access to storage is in `deployStatus`, converted to simple string and here storage is queried again.

Please take the logic from `deployStatus` here or prepare more structured data there and use storage access only once.

Reply:
The disadvantage of the previous implementation was string parsing. I got away from this by [adding a trait and derived classes](https://github.com/rchain/rchain/pull/3695/commits/f2682ce89b8c9d22de7d3a599c0e2cd2c5c8b3b1). Also, an unnecessary call to `BlockDagStorage` is not performed, and the hash of the block corresponding to the deployment is passed to the field of derived classes. I'm sure the code could be written more beautifully, but I couldn't figure out how.